Test all language generators in one command

jhowlett-scottlogic opened the pull request:
PR related to issue  https://github.com/ScottLogic/openapi-forge/issues/10

PR related to PR https://github.com/ScottLogic/openapi-forge-typescript/pull/8

Using openapi-forge test-generators you can test all language generators and then have a table of results presented. 

This command searches for a local language generators or uses the version hosted on GitHub.

jhowlett-scottlogic commented:
This approach uses a search and replace in some files (2 for TypeScript & 1 for CSharp) that refererences openapi-forge in the node_modules folder.  Instead these file paths are repalced to the local version of the generator and then reverted back. 

An alternative approach would be to have a separate configuration file in each language generator that would house these file paths (and another other variable data) and then the language generators will have a different configuration file when using the new test-generators  command compared to using the testing that is in each of the language generators.

A small refactoring to help prevent this testing from breaking if the file path lines change causing the regex patterns to start failing. 

Let me know what you think and i can bring this change in.

ColinEberhardt commented on README.md:
```diff
@@ -119,4 +119,26 @@ The following is a very high-level overview of the generation process:
+You can test openapi-forge on all of the language generators from one command:
+
+~~~
+% openapi-forge help test-generators
```
Adding as a new command for the CLI is a great idea 👍

ColinEberhardt commented on README.md:
```diff
@@ -119,4 +119,26 @@ The following is a very high-level overview of the generation process:
+Options:
+  -g, --generators <gens>            Narrow down the generators to test. Each letter is a generator, combine letters to test multiple generators, options are:
+                                     c (CSharp), t (TypeScript) (default: "ct")
+  -c, --csharp <csharpPath>          Sets the location of the CSharp generator. Default is a directory called 'openapi-forge-csharp' in the same location as
```
_aside_ should the forge have hard-coded references to the generators?

ColinEberhardt commented on src/generate.js:
```diff
@@ -172,14 +162,14 @@ async function generate(schemaPathOrUrl, generatorPathOrUrl, options) {
     }
 
-    if (schema && schema.components && schema.components.schemas) {
+    if (schema?.components?.schemas) {
```
much cleaner

ColinEberhardt commented on src/testGenerators.js:
```diff
@@ -0,0 +1,205 @@
+
+
+async function testGenerators(options) {
+    let temporaryFolder;
```
having lots of mutable variables is a bit of a 'code smell'!

jhowlett-scottlogic replied:
increased immutability

ColinEberhardt commented on src/testGenerators.js:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    shell.cd(__dirname, shellOptions);
+    if(typescript) {
```
From a 30,000 foot view I can see that code of a similar shape is duplicated later in this function. It is also quite long, and as more language generators are added will need refactoring. A sign that it is worth investing some more time in re-structuring.

jhowlett-scottlogic replied:
I have extracted duplicated logic to separate functions, generatorResolver.js or testResultParser.js

ColinEberhardt commented on src/testGenerators.js:
```diff
@@ -0,0 +1,205 @@
+            log.standard(`Loading TypeScript generator from '${generatorPath}'`);
+
+            // Check for local Typescript generator.
+            if (!fs.existsSync(generatorPath)) {
```
idea 💡 - this resolution process is something that is used twice in this function, but also, the generator itself has to perform this resolution. Could this be abstracted into a 'resolver', you give it a generator reference, it takes care of going away and finding it.

jhowlett-scottlogic replied:
done

ColinEberhardt commented on src/testGenerators.js:
```diff
@@ -0,0 +1,205 @@
+                featurePath = path.relative(generatorPath, path.join(__dirname, "../features/*.feature")).replaceAll("\\", "/");
+                basePath = path.relative(path.join(generatorPath, "features/support"), path.join(__dirname, "../src/generate")).replaceAll("\\", "/");
+            }
+            log.verbose("Replacing path to .feature files");
```
I'm really not sure why we need to change file paths? this feels quite fragile.

jhowlett-scottlogic replied:
Now uses cl arguments to configure the generators

ColinEberhardt commented on src/testGenerators.js:
```diff
@@ -0,0 +1,205 @@
+            const stdoutLength = stdout.length;
+            
+            // Format the output of the testing.
+            let result = stdout[stdoutLength-2].match(/^(\d+)m(\d+)\.\d+s/);
```
regex is a super powerful tool, but hard to maintain. Please make sure that you add a brief comment that explains the regex.

jhowlett-scottlogic replied:
I have added a small commnet above Regex uses to help explain them

ColinEberhardt commented on src/testGenerators.js:
```diff
@@ -0,0 +1,205 @@
+            // Format the output of the testing.
+            let result = stdout[stdoutLength-2].match(/^(\d+)m(\d+)\.\d+s/);
+
+            if(result) {
```
It looks like this code is parsing the TypeScript / cuccumber output in order to locate the test results. All looks good, but perhaps refactor this into a seperate file? `TestResultsParser` perhaps?

jhowlett-scottlogic replied:
done

ColinEberhardt commented on src/testGenerators.js:
```diff
@@ -0,0 +1,205 @@
+                resultArray.TypeScript = new Result(parseInt(result[1]), parseInt(result[9]), 0, parseInt(result[6]), parseInt(result[3]), time);
+                }
+
+                if(!temporaryFolder) {
```
See comment about generator resolving - could that also have responsibility for maintaining paths?

jhowlett-scottlogic replied:
generatorResolver.js handles getting and disposing of generators

ColinEberhardt commented on src/testGenerators.js:
```diff
@@ -0,0 +1,205 @@
+            }
+            log.standard(`${log.divider}`);
+        } finally {
+            if (temporaryFolder) {
```
clean-up could also be a job for the resolver?

ColinEberhardt reviewed (changes requested):
functionally works very well, but could do with a bit more refactoring.

jhowlett-scottlogic commented:
Changes since previous review:

- Handle new cl arguments in generator's `npm test`
- Extract parsing test results to testResultParser.js
- Extract generator download ans cleanup to generatorResolver.js
- Extract repeated logic in testGenerators()
- House shellOptions in log.js for one unified shellOptions
